### ISSUE TITLE
Dynamic runtime cross-domain interoperability discovery

### ISSUE THREAD

**Issue by onthebreeze** [ausdigital/ausdigital-dcp#21]
One issue with the DCP and UBL document & process identifiers (even when cleaned up to be simple and unambiguous as defined in #20) is that there is an expectation that implementers will build not only to a shared document semantic (eg UBL2.1) but also to every jurisdiction/industry/process context that they wish to support.  This might (just) be OK within a single jurisdiction like AU DBC invoicing.  But becomes very problematic for global ledger software vendors and global organisations that would like to support cross border or cross context exchanges.  

In many cases it could be that a simple document like a general purpose invoice or tax receipt that meets UBL 2.1 document semantics could also pass validation in multiple customisation contexts (eg both parties share the same DCP "DocumentIdentifier".  But it's unlikely that any implementer would take the risk to actually send an invoice to a party with an unrecognised DCP "ProcessIdentifier" because a validation failure at the receiver end point would most likely be blamed on the sender ledger by the end user.  

Therefore we propose a means to dynamically test validation rules as part of the DCP endpoint discovery and interoperability assessment process:

* The authority that defines the context specific rules (eg DBC) hosts a DCL that accepts the DCP Identifiers (DocumentIdentifier and ProcessIdentifier) as NAPTR queries and returns a URL (eg to http://bill.testpoint.io/api/v0/validator service) that hosts a validation API.
* A GET to the validationURL returns a link to the specification document (eg http://ausdigital.org/specs/ausdigital-bill/1.0/).
* A POST to the validationURL with a clear text sample of the document to be sent will invoke a validation of structure, codes, and rules and will return a standard JsonAPI error message which represents either a success or failure (with error or warning details).

Notes:

* The effort for the standard authority to "host a DCL" is trivial because it's just a case of manually maintaining perhaps a few dozen NAPTR records in their own domain.
* The DCL response can return a list of validation API end points that have been accredited by the authority as correctly implementing validation rules.  Sending ledgers would usually pick the first and, if unavailable, then pick the second etc.
* If there are multiple validation API providers (eg various access point providers or test service providers in the local network then the authority can re-order the URLs int he NAPTR response to "share the load". 
* Validation APIs are public (ie not authenticated) and free.  Therefore there is a DDOS risk. Validation API service providers may choose to throttle access to any rate limit that matches their chosen financial limits.
* Validation requires the transmission of a clear text document (over SSL) so that validation logic can work.  All validation services MUST "process and forget" the document content.  
* Validation service providers SHOULD make their validation code open source so that it can be inspected for accuracy and security and re-used by anyone.
* sending ledgers MAY choose to send a representative but anonymised sample to the validation API instead of the actual document queued for transmission of the secure network via the TAP protocol.

In this way, cross domain interoperability is enhanced because

* sending ledgers do not need to build and maintain compliance with potentially 100's of context specific rule-sets at design time.  Instead they can test at runtime and, if ok, then send a message. ie : "oh, good, the end points supports UBL 2.0, let me see if I can pass their localisation rules.  Good, it works, I'll send the invoice".
* the transparency of local specifications and accessibility to validation services will start to drive consistency across localised implementations. ie "we need to specify some localisation rules for invoice exchange in {country/industry X}, lets see what they did over there in {country/industry Y}."

**Comment by monkeypants** [comment]
> The authority that defines the context specific rules (eg DBC) hosts a DCL that accepts the DCP Identifiers (DocumentIdentifier and ProcessIdentifier) as NAPTR queries and returns a URL (eg to http://bill.testpoint.io/api/v0/validator service) that hosts a validation API.

Done this way, the owner of the customisation (from a governance perspective, the authority) is responsible for hosting the validator and publishing it's address. The hosting responsibility seems fair enough, but how is an integrator supposed to discover the authority for new customisations?

For example, imagine Alice has a new customer in a country she has never done business with (Bob). Bob has a business identifier from a scheme that Alice recognises, so she knows how to resolve (DCL) his DCP, and then look up his integration endpoints. She wants to send him an invoice, which she knows as Document ID "UBL 2.1 invoice" (per #20), and her particular customised invoice that she has to send is Process ID "DBC send-invoice". She can see Bob does not have an endpoint for "DBC send-invoice", however he does have an endpoint for "XYZ send-invoice", so Alice needs to find out if her DBC send-invoice happens to be a valid XYZ send-invoice. So she needs to dynamically discover the address of the XYZ send-invoice validator, using the XYZ send-invoice URN.

Before now, Alice has never heard of XYZ send-invoice authority. It's nice to know that they are responsible for hosting a "XYZ send-invoice" validation service, if only she can dynamically discover it...

If the XYZ customisation authority is responsible for hosting the DCL for discovering the address of the validators for XYZ customizations (process IDs), Alice still needs to know how to discover the root domain of XYZ validation DCL. Is this supposed to be well known? Is it published somewhere by some international committee? How is that supposed to work?

If there is some official list of customisation authorities, with mappings to their process validator DCL services, it would be simpler (for Alice) for that list to be a single process validator DCL service. As you pointed out, there will only be a small number of NAPTR records (one for every process ID of every customisation), so the technical technical cost/burden of hosting them is very small. The governance cost/burden may be significantly higher, but basically no different to the cost of agreeing on "who gets to be an officially listed customisation". I suppose the way to minimise that cost would to be have a permissive policy, "if you really want to host a customisation, and you are publishing validators and appropriate documentation, then you can", perhaps with a small fee to minimise spam/DDOS/nuisance publishing.

With a central DCL for customised process IDs, Alice would:
 1. DCL/DCP probe her new pan-jurisdictional customer, Bob
 2. Discover from his DCP that Bob does not have an exact-matching Process ID endpoint for the document she wishes to send him.
 3. Discover that Bob does have a different dialect/customisation for the same ProcessID semantic document (new Process ID URN). Possibly multiple, in fact.
 4. Use the central customisation DCL to discover the address of the validator for the new Process ID URN (or, if bob has multiple URNs published, discover the validators for all of them).
 5. Validate the document she wishes to send Bob, against the (single or multiple) validators for the customisations that Bob is able to receive
 6. If Alice's document happens to be valid for any of Bob's equivalent semantic endpoints (albeit with different customisations), send it (and expect it to work).
 7. If Alice's document is not valid for any of Bob's equivalent semantic endpoints, Alice will have both list of validation errors  and access to appropriate technical documentation, so she will be able to make a well informed decision about how to proceed.

Without a centralised DCL for customised Process ID URNs, we have a problem between steps 3 and 4 above. What alternatives are there?

**Comment by koriaf** [comment]
I think it's well-known table with values.
As I a sender party:
* I have receiver participant id
* I get scheme from this ID
* I know that given scheme is managed by this DCL
* so I can know that information about the validations also in the place where I got the DCL hostname.

**Comment by onthebreeze** [comment]
Yes, I'm not sure I see the problem here.  Its just DNS so I dont need to know where to look.  I look in DNS.  Something like this:

* a DCP query discovers a "processIdentifier" with "scheme" = ausdigital.org and "value" = bill.rcti.1.0" or something similar.
* I make a NAPTR query to <value>:<scheme> so bill.rcti.1,0&#58;ausdigital.org or something like that
* the NAPTR response is https://bill.testpoint.io/api/v0/validator or something like that.

Of course this depends on the NAPTR lookup key being domain based.  But that's a reasonable assumption.  If a local authority when publishing a context specific ruleset WANTS that ruleset to be dynamically discoverable and to allow runtime validation tests THEN it will make sure that it's DCP identifiers are suitable for NAPTR queries.

or am I missing something?

**Comment by monkeypants** [comment]
> I make a NAPTR query to {value}.{scheme} so bill-rcti-v1.ausdigital.org (might need to escape characters that aren't URL friendly)

Ah, that answers it. If the scheme is a domain name (controlled by the authority) then you are right, there is no need for any register of schemes. I didn't realise there was this constraint on valid Process ID names (must be domain name of NAPTR root domain).

**Comment by onthebreeze** [comment]
it isnt a currently specified constraint.  But I guess I'm saying that it will sort of evolve into a constraint / best practice because this capability is a desirable thing for the localisation authority.

**Comment by monkeypants** [comment]
If it is a hard constraint then we have an adequate solution:
 * Process ID customisation URN **must** have scheme name that is valid DNS name
 * MD5 hash of Process ID URN, prefixed with `b-`, and prepended to Process ID customisation scheme name, **must** resolve via NAPTR query to URL of validation service
 * validation service **must** conform to REST/JSON validation API for documents of specified Process ID customisation

If it's *only* a guideline or best practice, it doesn't obviate the need for central registry of customisations or other mechanism for distributing well-known validation capabilities.

**Comment by monkeypants** [comment]
I really like this solution, what's the next step @onthebreeze ?
